crates/holochain/src/core/ribosome/host_fn.rs

```diff
@@ -1,32 +1,158 @@
-pub mod agent_info;
-pub mod call;
-pub mod call_remote;
-pub mod capability_claims;
-pub mod capability_grants;
-pub mod capability_info;
-pub mod create;
-pub mod create_link;
-pub mod debug;
-pub mod decrypt;
-pub mod delete;
-pub mod delete_link;
-pub mod emit_signal;
-pub mod encrypt;
-pub mod entry_type_properties;
-pub mod get;
-pub mod get_agent_activity;
-pub mod get_details;
-pub mod get_link_details;
-pub mod get_links;
-pub mod hash_entry;
-pub mod property;
-pub mod query;
-pub mod random_bytes;
-pub mod schedule;
-pub mod show_env;
-pub mod sign;
-pub mod sys_time;
-pub mod unreachable;
-pub mod update;
-pub mod verify_signature;
-pub mod zome_info;
+use super::{CallContext, RibosomeT};
+use holochain_zome_types::zome_io::{HostFnApiErrorPlaceholder, HostFnApiT};
+use std::sync::Arc;
+
+pub struct HostFnApi<Ribosome: RibosomeT> {
+    ribosome: Arc<Ribosome>,
+    call_context: Arc<CallContext>,
+}
+
+impl<Ribosome: RibosomeT> HostFnApi<Ribosome> {
+    pub fn new(ribosome: Arc<Ribosome>, call_context: Arc<CallContext>) -> Self {
+        Self {
+            ribosome,
+            call_context,
+        }
+    }
+}
+
+macro_rules! host_fn_api_impls {
+    ( $( fn $f:ident ( $input:ty ) -> $output:ty; )* ) => {
+        $(
+            pub(crate) mod $f;
+        )*
+
+        impl<Ribosome: RibosomeT> HostFnApiT for HostFnApi<Ribosome> {
+            $(
+                fn $f(&self, input: $input) -> Result<$output, HostFnApiErrorPlaceholder> {
+                    let out = $f::$f(
+                        self.ribosome.clone(),
+                        self.call_context.clone(),
+                        input.into()
+                    ).map(|r| r.into_inner());
+                    Ok(out.expect("TODO, handle after crate re-org [B-03640]"))
+                }
+            )*
+        }
+    };
+}
+
+// All host_fn_api_impls below rely on this import
+use holochain_zome_types as zt;
+
+host_fn_api_impls! {
+
+    // ------------------------------------------------------------------
+    // These definitions are copy-pasted from
+    // holochain_zome_types::zome_io
+    // TODO: is there a way to unhygienically import this code in both places?
```

Comment:
This file is an interesting one to look at. This is where the `HostFnApiT` trait is implemented, which is what couples the host fns with the zome_io types, so we will get compiler errors if the two go out of sync.

I am wondering if anyone has any ideas about this. The following contents of this macro call are copypasta from `holochain_zome_types::zome_io`. I designed those two macros to take the same syntax but to spit out different code. It would be lovely if there were a way to DRY this up by unhygienically plopping this syntax into both places, or generating it somehow. `import!()` didn't work here.